### ISSUE TITLE
How to use Full-Text Search with Trees?

### ISSUE THREAD

**Issue by delphit** [typeorm/typeorm#2083]
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

Hello, I have service which works with any query generated by query builder. To use I call getData() and pass params and query created before.

```
  public getOwnNetwork(user: User) {
    return this.userRepository.createDescendantsQueryBuilder('users', 'usersClosure', user)
      .andWhere('users.parent = :id', { id: user.id });
  }
  public async getTeamWithPagination(payload: PaginationDto, userId: number) {
    const entity = await this.findOne(userId);
    const query = await this.getOwnNetwork(entity);
    return this.paginationService.getData(payload, { type: 'users', query, searchFields: ['id','email']});
  }
```
```
@Component()
export class PaginationService {
  public async getData(data: PaginationData, config: PaginationConfig) {
    const { page, limit } = data;
    const offset = (page - 1) * limit;
    const [items, count] = await this.buildQuery(data, config)
      .skip(offset)
      .take(limit)
      .getManyAndCount();
    return { data: items, count, pages: Math.ceil(count / limit) };
  }

  private buildQuery(data: PaginationData, config: PaginationConfig): SelectQueryBuilder<any> {
    return this.getSorting(data, config, this.getConditions(data, config));
  }

  private getConditions(data: PaginationData, config: PaginationConfig): SelectQueryBuilder<any> {
    const { type, where } = config;
    if (data.search) {
      config.searchFields.forEach((field, index) => {
        config.query =
          index === 0
            ? config.query.andWhere(`${type}.${field}::text like :keyword`)
            : config.query.orWhere(`${type}.${field}::text like :keyword`);
      });
    }
    return data.search ? config.query.setParameter('keyword', `%${data.search}%`) : config.query;
  }

  private getSorting(
    data: PaginationData,
    config: PaginationConfig,
    query: SelectQueryBuilder<any>,
  ): SelectQueryBuilder<any> {
    const { type } = config;
    return query.orderBy(
      `${type}.${data.orderBy ? data.orderBy : 'id'}`,
      data.sort.toString() == 'true' ? 'DESC' : 'ASC',
    );
  }
} 
  ```

Service is responsible for sorting and pagination. I use it for big lists, grids. But the problem is, I need to search through all passed to function row columns and match row if at least one column has matched to given search string.
Now it returns rows which not included in the previous query result.  

Long story short, I need to filter result of the previous query with new where conditions. Can you give a tip, or example how to implement this?

**Comment by pleerock** [comment]
can you please simplify your request, code and explain what you want to do in a simple way?

**Comment by delphit** [comment]
@pleerock, Okay :)
I have three:
```
user1 - parent.id null
user2 - parent.id 1
user3 - parent.id 2
```
And when I want to get descendants of **user1** I use this query:
```
await this.userRepository.createDescendantsQueryBuilder('users', 'usersClosure', user)
      .andWhere('users.id != :id', { id: user.id })
      .getMany();
```
I get result:
```
user2 - parent.id 1
user3 - parent.id 2
```
But the problems is begin when i try use **orWhere** in query.
```
await this.userRepository.createDescendantsQueryBuilder('users', 'usersClosure', user)
      .andWhere('users.id != :id', { id: user.id })
      .andWhere(`users.email::text like :keyword`)
      .orWhere(`users.id::text like :keyword`)
      .setParameter('keyword', `%3%`)
      .getMany();
```
Result:
```
user13 - parent.id null
user23 - parent.id null
user33 - parent.id 5
etc.
```
 I got all users from DB. But it is expected that the search will be only on my descendants.

**Comment by pleerock** [comment]
try:

```ts
await this.userRepository.createDescendantsQueryBuilder('users', 'usersClosure', user)
      .andWhere('(users.id != :id' AND users.email::text like :keyword'), { id: user.id })
      .orWhere(`users.id::text like :keyword`)
      .setParameter('keyword', `%3%`)
      .getMany();
```

If it won't work for you then create your own query based on [createDescendantsQueryBuilder](https://github.com/typeorm/typeorm/blob/master/src/repository/TreeRepository.ts#L80) method code.
